File: sdk/src/main/scala/com/horizen/account/abi/ABIUtil.scala
package com.horizen.account.abi

import com.horizen.account.state.ExecutionFailedException
import com.horizen.account.utils.Secp256k1
import com.horizen.utils.BytesUtils
import org.web3j.utils.Numeric
import scorex.crypto.hash.Keccak256

import java.nio.charset.StandardCharsets
import java.util

object ABIUtil {

  val METHOD_CODE_LENGTH = 4

  @throws(classOf[ExecutionFailedException])
  def getArgumentsFromData(data: Array[Byte]): Array[Byte] = getSlice(data, METHOD_CODE_LENGTH, data.length)

  @throws(classOf[ExecutionFailedException])
  def getFunctionSignature(data: Array[Byte]): String = BytesUtils.toHexString(getSlice(data, 0, METHOD_CODE_LENGTH))

  def getABIMethodId(methodSig: String): String = Numeric.toHexStringNoPrefix(Keccak256.hash(methodSig.getBytes(StandardCharsets.UTF_8)).take(4))

  @throws(classOf[ExecutionFailedException])
  private def getSlice(data: Array[Byte], from: Int, to: Int) = {
    if (data.length < METHOD_CODE_LENGTH) {
      throw new ExecutionFailedException(s"Data length ${data.length} must be >= $METHOD_CODE_LENGTH")
    }
    util.Arrays.copyOfRange(data, from, to)
  }
}

File: sdk/src/main/java/com/horizen/account/abi/ABIUtil.java
package com.horizen.account.abi;

import com.horizen.account.state.ExecutionFailedException;
import com.horizen.utils.BytesUtils;
import org.web3j.utils.Numeric;
import sparkz.crypto.hash.Keccak256;

import java.util.Arrays;

public class ABIUtil {

    public static final int METHOD_ID_LENGTH = 4;

    public static byte[] getArgumentsFromData(byte[] data) throws ExecutionFailedException {
        return getSlice(data, METHOD_ID_LENGTH, data.length);
    }

    public static String getFunctionSignature(byte[] data) throws ExecutionFailedException {
        return BytesUtils.toHexString(getSlice(data, 0, METHOD_ID_LENGTH));
    }

    private static byte[] getSlice(byte[] data, int from, int to) throws ExecutionFailedException {
        if (data.length < METHOD_ID_LENGTH) {
            throw new ExecutionFailedException("Data length " + data.length + " must be >= " + METHOD_ID_LENGTH);
        }
        return Arrays.copyOfRange(data, from, to);
    }

    public static String getABIMethodId(String methodSig) {
        return Numeric.toHexStringNoPrefix(Arrays.copyOfRange(((byte[]) Keccak256.hash(methodSig.getBytes())), 0, METHOD_ID_LENGTH));
    }

}

